## Importing libraries and data set

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# To establish a path to 02 Data folder
path=r'C:\Users\Shelb\OneDrive\Documents\CF Coursework\Advanced Analytics and Dashboard Design\2022 WA State CAFV Project'

In [3]:
df_EV_title=pd.read_csv(os.path.join(path,'02 Data','Original Data','Electric Vehicle Title and Registration Activity.csv'))

## Data exploration and descriptive statistics

In [4]:
# To show the size of the data set
df_EV_title.shape

(633874, 32)

In [5]:
# To show column names
df_EV_title.columns

Index(['Clean Alternative Fuel Vehicle Type', 'VIN (1-10)', 'Model Year',
       'Make', 'Model', 'New or Used Vehicle', 'Sale Price',
       'DOL Transaction Date', 'Transaction Type', 'Transaction Year',
       'Electric Vehicle Fee Paid', 'County', 'City', 'Postal Code',
       'Electric Range', 'Base MSRP', '2015 HB 2778 Exemption Eligibility',
       'Sale Date', 'Vehicle Primary Use', 'State of Residence',
       'DOL Vehicle ID', 'Legislative District',
       '2019 HB 2042 Clean Alternative Fuel Vehicle (CAFV) Eligibility',
       'Meets 2019 HB 2042 Electric Range Requirement',
       'Meets 2019 HB 2042 Sale Date Requirement',
       'Meets 2019 HB 2042 Sale Price/Value Requirement', 'Odometer Reading',
       'Odometer Code', 'Transportation Electrification Fee Paid',
       'Hybrid Vehicle Electrification Fee Paid', '2020 Census Tract',
       'Electric Utility'],
      dtype='object')

##### -Column names will need to be reformated without spaces to work in Python

In [6]:
# To show data types for each column
df_EV_title.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 633874 entries, 0 to 633873
Data columns (total 32 columns):
 #   Column                                                          Non-Null Count   Dtype  
---  ------                                                          --------------   -----  
 0   Clean Alternative Fuel Vehicle Type                             633874 non-null  object 
 1   VIN (1-10)                                                      633874 non-null  object 
 2   Model Year                                                      633874 non-null  int64  
 3   Make                                                            633874 non-null  object 
 4   Model                                                           633835 non-null  object 
 5   New or Used Vehicle                                             633874 non-null  object 
 6   Sale Price                                                      633874 non-null  int64  
 7   DOL Transaction Date                  

In [7]:
# To show descriptive statistics
df_EV_title.describe()

Model Year    Sale Price  Transaction Year    Postal Code  \
count  633874.000000  6.338740e+05     633874.000000  633738.000000   
mean     2017.160292  9.946394e+03       2019.348770   98135.171885   
std         3.143883  3.086735e+04          2.524943    2651.645901   
min      1993.000000  0.000000e+00       2010.000000     745.000000   
25%      2015.000000  0.000000e+00       2018.000000   98047.000000   
50%      2017.000000  0.000000e+00       2020.000000   98117.000000   
75%      2020.000000  0.000000e+00       2021.000000   98360.000000   
max      2023.000000  1.231202e+07       2022.000000   99901.000000   

       Electric Range      Base MSRP  DOL Vehicle ID  Legislative District  \
count   633874.000000  633874.000000    6.338740e+05         631687.000000   
mean       103.809408    3387.680186    2.026291e+08             30.272035   
std         94.405468   14883.095252    1.040864e+08             14.703528   
min          0.000000       0.000000    4.000000e+00              1.000000   
25%         19.000000       0.000000    1.387666e+08             20.000000   
50%         75.000000       0.000000    1.864201e+08             34.000000   
75%        208.000000       0.000000    2.393366e+08             43.000000   
max        337.000000  845000.000000    4.792548e+08             49.000000   

       Odometer Reading  2020 Census Tract  
count     633874.000000       6.337450e+05  
mean        2989.021687       5.294849e+10  
std        11747.381800       1.886913e+09  
min            0.000000       1.045020e+09  
25%            0.000000       5.303301e+10  
50%            0.000000       5.303303e+10  
75%           15.000000       5.305307e+10  
max       962619.000000       7.211913e+10

##### -The data type for the Legislative District column should be changed from float64 to object as the information it holds is categorical and not quantitative.

In [8]:
# To remove the limit on how many columns python shows in df.head() and display the first 5 rows of the set
pd.set_option('display.max_columns', None)
df_EV_title.head()

Clean Alternative Fuel Vehicle Type  VIN (1-10)  Model Year       Make  \
0      Battery Electric Vehicle (BEV)  7FCTGAAA8N        2022     RIVIAN   
1      Battery Electric Vehicle (BEV)  5YJ3E1EBXN        2022      TESLA   
2      Battery Electric Vehicle (BEV)  5YJ3E1EBXN        2022      TESLA   
3      Battery Electric Vehicle (BEV)  1G1FX6S07H        2017  CHEVROLET   
4      Battery Electric Vehicle (BEV)  1G1FX6S07H        2017  CHEVROLET   

     Model New or Used Vehicle  Sale Price DOL Transaction Date  \
0      R1T                 New           0       August 16 2022   
1  Model 3                 New       70440      October 11 2022   
2  Model 3                 New           0      October 11 2022   
3  Bolt EV                Used       21219      October 15 2019   
4  Bolt EV                Used           0      October 15 2019   

        Transaction Type  Transaction Year Electric Vehicle Fee Paid County  \
0  Original Registration              2022                        No   King   
1         Original Title              2022            Not Applicable   King   
2  Original Registration              2022                        No   King   
3         Original Title              2019            Not Applicable  Lewis   
4  Original Registration              2019                        No  Lewis   

       City  Postal Code  Electric Range  Base MSRP  \
0   SEATTLE      98107.0               0          0   
1   SEATTLE      98107.0               0          0   
2   SEATTLE      98107.0               0          0   
3  CHEHALIS      98532.0             238          0   
4  CHEHALIS      98532.0             238          0   

                  2015 HB 2778 Exemption Eligibility          Sale Date  \
0  ERROR: No battery range; TRANSACTION NOT ELIGI...                NaN   
1  ERROR: No battery range; TRANSACTION NOT ELIGI...  September 09 2022   
2  ERROR: No battery range; TRANSACTION NOT ELIGI...                NaN   
3  TRANSACTION NOT ELIGIBLE: Not a new vehicle; T...  September 21 2019   
4  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...                NaN   

  Vehicle Primary Use State of Residence  DOL Vehicle ID  \
0               Truck                 WA       216581403   
1           Passenger                 WA       219017578   
2           Passenger                 WA       219017578   
3           Passenger                 WA       477229294   
4           Passenger                 WA       477229294   

   Legislative District  \
0                  43.0   
1                  43.0   
2                  43.0   
3                  20.0   
4                  20.0   

  2019 HB 2042 Clean Alternative Fuel Vehicle (CAFV) Eligibility  \
0  ERROR: No battery range; TRANSACTION NOT ELIGI...               
1  ERROR: No battery range; TRANSACTION NOT ELIGI...               
2  ERROR: No battery range; TRANSACTION NOT ELIGI...               
3                     Clean Alternative Fuel Vehicle               
4  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...               

   Meets 2019 HB 2042 Electric Range Requirement  \
0                                          False   
1                                          False   
2                                          False   
3                                           True   
4                                           True   

   Meets 2019 HB 2042 Sale Date Requirement  \
0                                     False   
1                                      True   
2                                     False   
3                                      True   
4                                     False   

   Meets 2019 HB 2042 Sale Price/Value Requirement  Odometer Reading  \
0                                            False                 0   
1                                            False                15   
2                                            False                 0   
3                                             True             43515

##### -The 2015 HB 2778 Exemption Eligibility and 2019 HB 2042 Clean Alternative Fuel Vehicle (CAFV) Eligibility columns contain both whether the vehicle is eligible or not, and in the event that it is not eligible the reasons why are listed. This should be split into separate columns to make the information useful for this analysis.

##### - The 2020 Census Tract column is not necessary for this analysis 

## Reformatting column names and dropping unnecessary column

In [9]:
# To remove spaces from column names for use in Python
df_EV_title.rename(columns={'Clean Alternative Fuel Vehicle Type':'Clean_Alternative _Fuel _Vehicle _Type'},inplace=True)
df_EV_title.rename(columns={'VIN (1-10)':'VIN'},inplace=True)
df_EV_title.rename(columns={'Model Year':'Model_Year'},inplace=True)
df_EV_title.rename(columns={'New or Used Vehicle':'New_or_Used_Vehicle'},inplace=True)
df_EV_title.rename(columns={'Sale Price':'Sale_Price'},inplace=True)
df_EV_title.rename(columns={'DOL Transaction Date':'DOL_Transaction_Date'},inplace=True)
df_EV_title.rename(columns={'Transaction Type':'Transaction_Type'},inplace=True)
df_EV_title.rename(columns={'Transaction Year':'Transaction_Year'},inplace=True)
df_EV_title.rename(columns={'Electric Vehicle Fee Paid':'Electric_Vehicle_Fee_Paid'},inplace=True)
df_EV_title.rename(columns={'Postal Code':'Postal_Code'},inplace=True)

In [10]:
df_EV_title.rename(columns={'Electric Range':'Electric_Range'},inplace=True)
df_EV_title.rename(columns={'Base MSRP':'Base_MSRP'},inplace=True)
df_EV_title.rename(columns={'2015 HB 2778 Exemption Eligibility':'2015_HB_2778_Exemption_Eligibility'},inplace=True)
df_EV_title.rename(columns={'Sale Date':'Sale_Date'},inplace=True)
df_EV_title.rename(columns={'Vehicle Primary Use':'Vehicle_Primary_Use'},inplace=True)
df_EV_title.rename(columns={'State of Residence':'State_of_Residence'},inplace=True)
df_EV_title.rename(columns={'DOL Vehicle ID ':'DOL_Vehicle_ID'},inplace=True)
df_EV_title.rename(columns={'Legislative District':'Legislative_District'},inplace=True)
df_EV_title.rename(columns={'2019 HB 2042 Clean Alternative Fuel Vehicle (CAFV) Eligibility':'2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility'},inplace=True)
df_EV_title.rename(columns={'Meets 2019 HB 2042 Electric Range Requirement':'Meets_2019_HB_2042_Electric_Range_Requirement'},inplace=True)

In [11]:
df_EV_title.rename(columns={'Meets 2019 HB 2042 Sale Date Requirement':'Meets_2019_HB_2042_Sale_Date_Requirement'},inplace=True)
df_EV_title.rename(columns={'Meets 2019 HB 2042 Sale Price/Value Requirement':'Meets_2019_HB_2042_Sale_Price/Value_Requirement'},inplace=True)
df_EV_title.rename(columns={'Odometer Reading':'Odometer_Reading'},inplace=True)
df_EV_title.rename(columns={'Odometer Code':'Odometer_Code'},inplace=True)
df_EV_title.rename(columns={'Transportation Electrification Fee Paid':'Transportation_Electrification_Fee_Paid'},inplace=True)
df_EV_title.rename(columns={'Hybrid Vehicle Electrification Fee Paid':'Hybrid_Vehicle_Electrification_Fee_Paid'},inplace=True)
df_EV_title.rename(columns={'2020 Census Tract':'2020_Census_Tract'},inplace=True)
df_EV_title.rename(columns={'Electric Utility':'Electric_Utility'},inplace=True)

In [12]:
# To confirm successful column name changes
df_EV_title.columns

Index(['Clean_Alternative _Fuel _Vehicle _Type', 'VIN', 'Model_Year', 'Make',
       'Model', 'New_or_Used_Vehicle', 'Sale_Price', 'DOL_Transaction_Date',
       'Transaction_Type', 'Transaction_Year', 'Electric_Vehicle_Fee_Paid',
       'County', 'City', 'Postal_Code', 'Electric_Range', 'Base_MSRP',
       '2015_HB_2778_Exemption_Eligibility', 'Sale_Date',
       'Vehicle_Primary_Use', 'State_of_Residence', 'DOL Vehicle ID',
       'Legislative_District',
       '2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility',
       'Meets_2019_HB_2042_Electric_Range_Requirement',
       'Meets_2019_HB_2042_Sale_Date_Requirement',
       'Meets_2019_HB_2042_Sale_Price/Value_Requirement', 'Odometer_Reading',
       'Odometer_Code', 'Transportation_Electrification_Fee_Paid',
       'Hybrid_Vehicle_Electrification_Fee_Paid', '2020_Census_Tract',
       'Electric_Utility'],
      dtype='object')

In [13]:
# To drop unnecessary 2020_Census_Tract column
df_EV_title=df_EV_title.drop(columns = ['2020_Census_Tract'])

## Fixing columns with mixed data types

In [14]:
# To check for columns with mixed data types
for col in df_EV_title.columns.tolist():
  weird = (df_EV_title[[col]].applymap(type) != df_EV_title[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_EV_title[weird]) > 0:
    print (col)

Model
County
City
Sale_Date
State_of_Residence
2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility
Transportation_Electrification_Fee_Paid
Hybrid_Vehicle_Electrification_Fee_Paid
Electric_Utility


In [15]:
# To change mixed data types to string
df_EV_title['Model']=df_EV_title['Model'].astype('str')
df_EV_title['County']=df_EV_title['County'].astype('str')
df_EV_title['City']=df_EV_title['City'].astype('str')
df_EV_title['Sale_Date']=df_EV_title['Sale_Date'].astype('str')
df_EV_title['State_of_Residence']=df_EV_title['State_of_Residence'].astype('str')
df_EV_title['2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility']=df_EV_title['2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility'].astype('str')
df_EV_title['Transportation_Electrification_Fee_Paid']=df_EV_title['Transportation_Electrification_Fee_Paid'].astype('str')
df_EV_title['Hybrid_Vehicle_Electrification_Fee_Paid']=df_EV_title['Hybrid_Vehicle_Electrification_Fee_Paid'].astype('str')
df_EV_title['Electric_Utility']=df_EV_title['Electric_Utility'].astype('str')

## Finding and addressing missing values

In [16]:
# To find NaN missing values
df_EV_title.isnull().sum()

Clean_Alternative _Fuel _Vehicle _Type                               0
VIN                                                                  0
Model_Year                                                           0
Make                                                                 0
Model                                                                0
New_or_Used_Vehicle                                                  0
Sale_Price                                                           0
DOL_Transaction_Date                                                 0
Transaction_Type                                                     0
Transaction_Year                                                     0
Electric_Vehicle_Fee_Paid                                            0
County                                                               0
City                                                                 0
Postal_Code                                                        136
Electr

##### There are 136 missing values in the Postal_Code column and 2187 missing values in the Legislative_District column

In [17]:
# To create a subset for NaN values in Postal_Code
df_nan_post_code=df_EV_title[df_EV_title['Postal_Code'].isnull()==True]
df_nan_post_code

Clean_Alternative _Fuel _Vehicle _Type         VIN  Model_Year  \
11936   Plug-in Hybrid Electric Vehicle (PHEV)  1G1RB6E47F        2015   
11937   Plug-in Hybrid Electric Vehicle (PHEV)  1G1RB6E47F        2015   
11938   Plug-in Hybrid Electric Vehicle (PHEV)  1G1RB6E47F        2015   
11939   Plug-in Hybrid Electric Vehicle (PHEV)  1G1RB6E47F        2015   
55909           Battery Electric Vehicle (BEV)  5YJSA1E42G        2016   
...                                        ...         ...         ...   
606816  Plug-in Hybrid Electric Vehicle (PHEV)  1G1RD6E4XE        2014   
612848          Battery Electric Vehicle (BEV)  5YJXCBE27J        2018   
613487  Plug-in Hybrid Electric Vehicle (PHEV)  5UXKT0C38H        2017   
614652          Battery Electric Vehicle (BEV)  5YJSA1E24G        2016   
614653          Battery Electric Vehicle (BEV)  5YJSA1E24G        2016   

             Make    Model New_or_Used_Vehicle  Sale_Price  \
11936   CHEVROLET     Volt                Used           0   
11937   CHEVROLET     Volt                Used           0   
11938   CHEVROLET     Volt                Used           0   
11939   CHEVROLET     Volt                Used           0   
55909       TESLA  Model S                Used           0   
...           ...      ...                 ...         ...   
606816  CHEVROLET     Volt                Used           0   
612848      TESLA  Model X                 New           0   
613487        BMW       X5                Used       55000   
614652      TESLA  Model S                Used           0   
614653      TESLA  Model S                 New           0   

       DOL_Transaction_Date       Transaction_Type  Transaction_Year  \
11936      December 07 2018  Original Registration              2018   
11937      December 30 2019   Registration Renewal              2019   
11938      December 21 2020   Registration Renewal              2020   
11939      December 28 2021   Registration Renewal              2021   
55909         March 03 2020  Original Registration              2020   
...                     ...                    ...               ...   
606816        April 25 2022   Registration Renewal              2022   
612848     November 07 2018  Original Registration              2018   
613487     February 28 2018         Transfer Title              2018   
614652      January 29 2018   Registration Renewal              2018   
614653      January 04 2017  Original Registration              2017   

       Electric_Vehicle_Fee_Paid County City  Postal_Code  Electric_Range  \
11936                         No    nan  nan          NaN              38   
11937                        Yes    nan  nan          NaN              38   
11938                        Yes    nan  nan          NaN              38   
11939                        Yes    nan  nan          NaN              38   
55909                         No    nan  nan          NaN             210   
...                          ...    ...  ...          ...             ...   
606816                       Yes    nan  nan          NaN              38   
612848                        No    nan  nan          NaN             238   
613487            Not Applicable    nan  nan          NaN              14   
614652                       Yes    nan  nan          NaN             210   
614653                        No    nan  nan          NaN             210   

        Base_MSRP                 2015_HB_2778_Exemption_Eligibility  \
11936           0  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...   
11937           0  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...   
11938           0  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...   
11939           0  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...   
55909           0  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...   
...           ...                                                ...   
606816          0  TRANSACTION NOT ELIGIBLE: Non-sale, registrati...   
612848  

##### A number of records missing Postal_Code values are also missing city, county, and legislative district information. Many of these records are for EV buyers who reside outside of Washington state

In [18]:
# To look for records of buyers outside of Washington state
df_EV_title['State_of_Residence'].value_counts(dropna=False)

WA     631688
CA        490
VA        227
TX        157
OR        153
MD        118
FL         90
AZ         74
NV         59
GA         59
CO         51
IL         50
NY         49
DC         38
NC         37
HI         35
CT         33
MA         32
MI         32
AK         29
ID         27
PA         27
NJ         27
MN         25
SC         21
UT         21
TN         18
OK         17
IN         16
MO         15
MT         14
NE         14
OH         12
LA         11
AL         11
NM         10
KS         10
NH         10
AR          8
KY          7
BC          6
MS          6
RI          6
WY          5
WI          4
ME          4
ND          3
DE          3
XX          3
AE          3
PR          3
SD          2
ON          2
QC          1
nan         1
Name: State_of_Residence, dtype: int64

In [19]:
# To create a flag for out of state buyers
result=[]

for value in df_EV_title['State_of_Residence']:
    if value == 'WA':
        result.append('Local_Resident')
    else:
        result.append('Out_of_State_Resident')

In [20]:
df_EV_title['Out_of_State_Buyer?']=result
df_EV_title['Out_of_State_Buyer?'].value_counts(dropna=False)

Local_Resident           631688
Out_of_State_Resident      2186
Name: Out_of_State_Buyer?, dtype: int64

##### The number of out of state buyers accounts for both the missing values in the Postal_Code and Legislative_District columns

## Finding and addressing duplicates

In [21]:
# To find duplicate records
df_duplicates=df_EV_title[df_EV_title.duplicated()]

In [22]:
df_duplicates

Clean_Alternative _Fuel _Vehicle _Type         VIN  Model_Year  \
13958           Battery Electric Vehicle (BEV)  5YJSA1E27F        2015   
52202           Battery Electric Vehicle (BEV)  5YJ3E1EB9L        2020   
82296           Battery Electric Vehicle (BEV)  537LS7D49B        2011   
112059          Battery Electric Vehicle (BEV)  5YJ3E1EB9K        2019   
222591          Battery Electric Vehicle (BEV)  5YJSA1H12E        2014   
269732  Plug-in Hybrid Electric Vehicle (PHEV)  WBY1Z8C52H        2017   
272643  Plug-in Hybrid Electric Vehicle (PHEV)  SALWR4RY9M        2021   
306932          Battery Electric Vehicle (BEV)  1G1FX6S0XJ        2018   
328281          Battery Electric Vehicle (BEV)  1N4BZ0CP3H        2017   
486270          Battery Electric Vehicle (BEV)  1N4AZ0CP6F        2015   
562563  Plug-in Hybrid Electric Vehicle (PHEV)  LYVBR0DL8K        2019   
590120  Plug-in Hybrid Electric Vehicle (PHEV)  JTDKN3DP3C        2012   

                  Make                     Model New_or_Used_Vehicle  \
13958            TESLA                   Model S                Used   
52202            TESLA                   Model 3                Used   
82296   AZURE DYNAMICS  Transit Connect Electric                Used   
112059           TESLA                   Model 3                Used   
222591           TESLA                   Model S                Used   
269732             BMW                        i3                Used   
272643      LAND ROVER         Range Rover Sport                Used   
306932       CHEVROLET                   Bolt EV                Used   
328281          NISSAN                      Leaf                Used   
486270          NISSAN                      Leaf                Used   
562563           VOLVO                      XC60                Used   
590120          TOYOTA             Prius Plug-in                Used   

        Sale_Price DOL_Transaction_Date      Transaction_Type  \
13958            0         July 21 2022  Registration Renewal   
52202            0     December 15 2021  Registration Renewal   
82296            0        April 02 2014  Registration Renewal   
112059           0    September 09 2022  Registration Renewal   
222591           0    September 04 2021  Registration Renewal   
269732           0         July 14 2018  Registration Renewal   
272643           0    September 23 2022  Registration Renewal   
306932           0        March 10 2021  Registration Renewal   
328281           0       August 21 2020  Registration Renewal   
486270           0      October 05 2017  Registration Renewal   
562563           0     December 18 2020  Registration Renewal   
590120           0         July 06 2021  Registration Renewal   

        Transaction_Year Electric_Vehicle_Fee_Paid        County  \
13958               2022                       Yes          King   
52202               2021                       Yes          King   
82296               2014                       Yes          King   
112059              2022                       Yes       Spokane   
222591              2021                       Yes          King   
269732              2018                       Yes          King   
272643              2022                        No          King   
306932              2021                       Yes      Thurston   
328281              2020                       Yes          King   
486270              2017                       Yes          King   
562563              2020                        No  Grays Harbor   
590120              2021                        No          King   

                City  Postal_Code  Electric_Range  Base_MSRP  \
13958      SAMMAMISH      98074.0             208          0   
52202       ISSAQUAH      98029.0             322          0   
82296        SEATTLE      98107.0              56          0   
112059        CHENEY      99004.0             220          0   
222591       SEATTLE      98117.0             208      69

In [23]:
# To check the shape of df_duplicates
df_duplicates.shape

(12, 32)

##### There are 12 duplicate records

In [24]:
# To check the shape of df_EV_title
df_EV_title.shape

(633874, 32)

In [25]:
# To remove duplicates from the data set
df_EV_title_clean=df_EV_title.drop_duplicates()

In [26]:
# To check the new shape of the data set
df_EV_title_clean.shape

(633862, 32)

##### Duplicates successfully removed

## To address the useability 2015_HB_2778_Exemption_Eligibility and 2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility columns

In [27]:
# To count the values in 2015_HB_2778_Exemption_Eligibility
df_EV_title_clean['2015_HB_2778_Exemption_Eligibility'].value_counts()

TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction                                                                                                                                                           335810
VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction                                                                                                             70974
ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction                                                                                                                                   57677
ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined                                                                                39805
TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined 

##### Clean Alternative Fuel Vehicle indicates eligibility for exemption
##### There seem to be three categories for non eligibility. Reasons for non eligibility listed are as follows: 

#### Transaction Not Eligible
##### --Non sale, registration transaction
##### --Sale before 7/15/2015
##### --Not a new vehicle
##### --Sale after 5/31/2018

#### Vehicle Model Not Eligible
##### --Low battery range
##### --Base MSRP too high

#### Cannot Determine Eligibility
##### --Base MSRP not determined
##### --No battery range

In [28]:
# To create a simplified column flagging 2015 HB 2778 eligibility
result=[]

for value in df_EV_title_clean['2015_HB_2778_Exemption_Eligibility']:
    if value == 'Clean Alternative Fuel Vehicle':
        result.append('Eligible')
    else:
        result.append('Not Eligible')

In [29]:
df_EV_title_clean['2015_HB_2778_Eligible?']=result

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\3396968791.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Eligible?']=result


In [30]:
df_EV_title_clean['2015_HB_2778_Eligible?'].value_counts()

Not Eligible    632948
Eligible           914
Name: 2015_HB_2778_Eligible?, dtype: int64

In [49]:
# To change 2015_HB_2778_Eligible? to string data type
df_EV_title_clean['2015_HB_2778_Eligible?']=df_EV_title_clean['2015_HB_2778_Eligible?'].astype('str')

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\3346023776.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Eligible?']=df_EV_title_clean['2015_HB_2778_Eligible?'].astype('str')


In [50]:
# To create a column for 2015 HB 2778 Transaction Eligibility
result=[]

for value in df_EV_title_clean['2015_HB_2778_Exemption_Eligibility']:
    if value == 'TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction':
        result.append('Transaction Not Eligible')
    elif value == 'ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction':
        result.append('Transaction Not Eligible')
    elif value == 'ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle':
        result.append('Transaction Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Transaction Not Eligible')
    else:
        result.append('Eligible Transaction')

In [51]:
df_EV_title_clean['2015_HB_2778_Transaction_Eligible?']=result

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\3137136193.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Transaction_Eligible?']=result


In [52]:
df_EV_title_clean['2015_HB_2778_Transaction_Eligible?'].value_counts()

Transaction Not Eligible    614475
Eligible Transaction         19387
Name: 2015_HB_2778_Transaction_Eligible?, dtype: int64

In [53]:
# To change 2015_HB_2778_Transaction_Eligible? to string data type
df_EV_title_clean['2015_HB_2778_Transaction_Eligible?']=df_EV_title_clean['2015_HB_2778_Transaction_Eligible?'].astype('str')

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\3260150776.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Transaction_Eligible?']=df_EV_title_clean['2015_HB_2778_Transaction_Eligible?'].astype('str')


In [54]:
# To create a column for 2015 HB 2778 Vehicle Model Eligibility
result=[]

for value in df_EV_title_clean['2015_HB_2778_Exemption_Eligibility']:
    if value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Model Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018':
        result.append('Model Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle':
        result.append('Model Not Eligible')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; VEHICLE MODEL NOT ELIGIBLE: Base MSRP is too high':
        result.append('Model Not Eligible')
    else:
        result.append('Eligible Vehicle Model')

In [55]:
df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?']=result

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\502675508.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?']=result


In [56]:
df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?'].value_counts()

Eligible Vehicle Model    535261
Model Not Eligible         98601
Name: 2015_HB_2778_Vehicle_Model_Eligibile?, dtype: int64

In [57]:
df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?']=df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?'].astype('str')

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\2396373170.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?']=df_EV_title_clean['2015_HB_2778_Vehicle_Model_Eligibile?'].astype('str')


In [58]:
# To create a column for 2015 HB 2778 Missing Vehicle Information
result=[]

for value in df_EV_title_clean['2015_HB_2778_Exemption_Eligibility']:
    if value == 'ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined ':
        result.append('Cannot Determine Eligibility')
    elif value == 'TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Sale after 5/31/2018; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    elif value == 'TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined ':
        result.append('Cannot Determine Eligibility')
    elif value == 'VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Not a new vehicle; TRANSACTION NOT ELIGIBLE: Transaction before 7/15/2015; CANNOT DETERMINE ELIGIBILITY: Base MSRP has not been determined':
        result.append('Cannot Determine Eligibility')
    else:
        result.append('All Vehicle Information Present')

In [59]:
df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information']= result

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\3819807178.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information']= result


In [60]:
df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information'].value_counts()

All Vehicle Information Present    513733
Cannot Determine Eligibility       120129
Name: 2015_HB_2778_Missing_Vehicle_Information, dtype: int64

In [61]:
df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information']=df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information'].astype('str')

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\1043441379.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information']=df_EV_title_clean['2015_HB_2778_Missing_Vehicle_Information'].astype('str')


In [63]:
# To show value counts for the 2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility column
pd.set_option('display.max_rows', None)
df_EV_title_clean['2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility'].value_counts()

TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction                                                                                                                                                               335810
VEHICLE MODEL NOT ELIGIBLE: Low battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction                                                                                                                 70974
ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Non-sale, registration transaction                                                                                                                                       57677
ERROR: No battery range; TRANSACTION NOT ELIGIBLE: Sale price too high                                                                                                                                                      34209
TRANSACTION NOT ELIGIBLE: Sale before August 2019; CAUTION: Sale price not available            

In [64]:
# To create a simplified column flagging 2019 HB 2042 Eligibility
result=[]

for value in df_EV_title_clean['2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility']:
    if value == 'Clean Alternative Fuel Vehicle':
        result.append('Eligible')
    else:
        result.append('Not Eligible')

In [66]:
df_EV_title_clean['2019_HB_2042_Eligible?']=result

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\2774926025.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2019_HB_2042_Eligible?']=result


In [67]:
df_EV_title_clean['2019_HB_2042_Eligible?'].value_counts()

Not Eligible    612445
Eligible         21417
Name: 2019_HB_2042_Eligible?, dtype: int64

In [68]:
# To change 2019_HB_2042_Eligible? to string data type
df_EV_title_clean['2019_HB_2042_Eligible?']=df_EV_title_clean['2019_HB_2042_Eligible?'].astype('str')

C:\Users\Shelb\AppData\Local\Temp\ipykernel_19424\722584305.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_EV_title_clean['2019_HB_2042_Eligible?']=df_EV_title_clean['2019_HB_2042_Eligible?'].astype('str')


In [70]:
# To remove the unusable 2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility and 2015_HB_2778_Exemption_Eligibility from the cleaned data set
df_EV_title_clean=df_EV_title_clean.drop(columns = ['2019_HB_2042_Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility','2015_HB_2778_Exemption_Eligibility'])

## Exporting and saving

In [71]:
# To check the shape of the cleaned data set
df_EV_title_clean.shape

(633862, 35)

In [72]:
# To save and export as a CSV file
df_EV_title_clean.to_csv(os.path.join(path,'02 Data','Prepared Data','clean_EV_title_and_reg_activity_12_2022.csv'))